FreeOrion fails to build on archs other that amd64

Bug Report
==========
<!--
Please fill in a a meaningful and concise title in the field above.
It helps us to classify the issue even  before reading the body and handling the
issue a bit better.

Examples:
  * Great: "Unable to select fleet in galaxy map"
  * Good: "Unable to click fleet"
  * Not really okay: "Issue with fleets"
  * Horrible: "HELP!!!"
-->


Environment
-----------
<!--
You don't need to provide the information in section when they are not useful in
the context of this issue.  For example a crashing hosted server may not need
information about the graphic card, but maybe depend on the operating system you
use.  When in doubt provide all information you have available.

You can find the FreeOrion version number in the lower right corner of the game
main screen.

When stating the operating system also note the version of the operating system
so, `Windows 8 Pro` instead of `Windows` or `Mac OSX Mountain Lion` instead of
`OSX`.
-->

* **FreeOrion Version**:  Weekly test build 2023-06-27.1b21f9b
* **Operating System**: Linux Ubuntu
* **Graphic card used**:
* **Fetched as** <!-- delete lines below that don't apply -->
  * Compiled from source


Description
-----------
<!--
Add a meaningful description of the bug you encountered.

If a screenshot or image helps to describe the issue content feel free to create
one.

Also attach the log files the game creates.  The log files The log files can be
found on
* Windows: %APPDATA%\FreeOrion
* MacOSX: ~/Library/"Application Support"/FreeOrion
* Linux: ${XDG_DATA_HOME:-~/.local/share}/freeorion
-->

When trying to build PPA for Ubuntu I got compilation errors for different archs:

```
/<<PKGBUILDDIR>>/util/CheckSums.cpp:28:28: error: static assertion failed
   28 |     static_assert(char_cs2 == 61);
      |                   ~~~~~~~~~^~~~~
/<<PKGBUILDDIR>>/util/CheckSums.cpp:28:28: note: the comparison reduces to ‘(195 == 61)’
/<<PKGBUILDDIR>>/util/CheckSums.cpp:29:28: error: static assertion failed
   29 |     static_assert(char_cs3 == 15);
      |                   ~~~~~~~~~^~~~~
/<<PKGBUILDDIR>>/util/CheckSums.cpp:29:28: note: the comparison reduces to ‘(241 == 15)’
make[3]: *** [CMakeFiles/freeorioncommon.dir/build.make:779: CMakeFiles/freeorioncommon.dir/util/CheckSums.cpp.o] Error 1
```

https://launchpadlibrarian.net/674319701/buildlog_ubuntu-mantic-arm64.freeorion_0.5.999+1SNAPSHOT20230627ppa1~mantic_BUILDING.txt.gz

Expected Result
---------------
<!--
When you enter an issue please add a description of what behaviour you would
expect from the game instead of the issue.
-->

Successful build.

Steps to reproduce
------------------
<!--
If the error only manifests itself after doing a certain number of actions
please add a list of steps to reproduce the bug.  If the error only occurs in
a certain game state please be sure to add a save game to the issue.

To upload save games please pack them into a Zip archive (*.zip) first.
-->
* Build FreeOrion for Linux on non-amd64 arch.

--- a/util/CheckSums.cpp
+++ b/util/CheckSums.cpp
@@ -1,33 +1,99 @@
-#include "CheckSums.h"
+﻿#include "CheckSums.h"
 
+#include <array>
+#if __has_include(<bit>)
+#  include <bit>
+#endif
 #include <cassert>
 #include <cmath>
 #include <float.h>
 #include <limits>
 
 namespace CheckSums {
-    static constexpr auto char_cs1 = []() {
-        uint32_t sum = 0;
-        CheckSumCombine(sum, 'q');
-        return sum;
-    }();
-    static constexpr auto char_cs2 = []() {
-        uint32_t sum = 0;
-        CheckSumCombine(sum, static_cast<char>(195));
-        return sum;
-    }();
-    static constexpr auto char_cs3 = []() {
-        uint32_t sum = 0;
-        CheckSumCombine(sum, static_cast<char>(-15));
-        return sum;
-    }();
-    static_assert(char_cs1 == 'q');
+    auto csc_ = [](const auto in, const auto csc_) noexcept {
+        if constexpr (std::is_arithmetic_v<decltype(in)>) {
+            uint32_t sum = 0;
+            CheckSumCombine(sum, in);
+            return sum;
+        } else {
+            std::array<uint32_t, in.size()> sums = {0};
+            for (size_t idx = 0; idx < sums.size(); ++idx)
+                sums[idx] = csc_(in[idx], csc_);
+            return sums;
+        }
+    };
+    auto csc = [](const auto in) noexcept { return csc_(in, csc_); };
+    static_assert(csc('q') == 'q');
 
-#if !defined(FREEORION_ANDROID)
-    // TODO: figure out why these fail on Android builds...
-    static_assert(char_cs2 == 61);
-    static_assert(char_cs3 == 15);
+    static_assert(CHAR_BIT == 8);
+    constexpr auto ucvals = std::array<unsigned char, 15>{0,1,40,'q',120,126,127,128,129,130,200,250,253,254,255};
+    constexpr auto scvals = std::array<signed char, 15>  {0,1,40,'q',120,126,127,-128,-127,-126,-56,-6,-3,-2,-1};
+#if defined(__cpp_lib_bit_cast)
+    static_assert(std::bit_cast<decltype(ucvals)>(scvals) == ucvals);
+    static_assert(std::bit_cast<decltype(scvals)>(ucvals) == scvals);
+    static_assert(std::bit_cast<std::array<char, ucvals.size()>>(ucvals) ==
+                  std::bit_cast<std::array<char, scvals.size()>>(scvals));
 #endif
+    static_assert(std::is_convertible_v<unsigned char, char> || std::is_convertible_v<signed char, char>);
+    static_assert(std::is_convertible_v<char, unsigned char> || std::is_convertible_v<char, signed char>);
+
+    constexpr auto uc2scvals = [](const auto ucvals){
+        std::array<signed char, ucvals.size()> uc2scvals{};
+        for (size_t idx = 0u; idx < ucvals.size(); ++idx)
+            uc2scvals[idx] = static_cast<signed char>(ucvals[idx]);
+        return uc2scvals;
+    }(ucvals);
+
+    constexpr auto sc2ucvals = [](const auto scvals) {
+        std::array<unsigned char, scvals.size()> sc2ucvals{};
+        for (size_t idx = 0u; idx < scvals.size(); ++idx)
+            sc2ucvals[idx] = static_cast<unsigned char>(scvals[idx]);
+        return sc2ucvals;
+    }(scvals);
+
+    static constexpr auto csc1 = csc(ucvals);
+    static constexpr auto csc2 = csc(sc2ucvals);
+    static constexpr auto csc3 = csc(uc2scvals);
+    static constexpr auto csc4 = csc(scvals);
+    static_assert(csc1 == csc2);
+    static_assert(csc2 == csc3);
+    static_assert(csc3 == csc4);
+    static_assert(csc1[3] == 'q');
+    static_assert([]() {
+                    for (size_t idx = 0u; idx < csc1.size(); ++idx)
+                        if (csc1[idx] != ucvals[idx])
+                            return false;
+                    return true;
+                  }());
+
+#if defined(__cpp_lib_char8_t)
+    constexpr std::u8string_view long_chars = u8"αbåオーガитیای مجهو ";
+    constexpr auto long_chars_arr = []() {
+        std::array<std::string_view::value_type, long_chars.size()> retval{};
+        for (std::size_t idx = 0; idx < retval.size(); ++idx)
+            retval[idx] = long_chars[idx];
+        return retval;
+    }();
+    constexpr std::string_view long_chars_sv(long_chars_arr.data(), long_chars_arr.size());
+#else
+    constexpr std::string_view long_chars_sv = u8"αbåオーガитیای مجهو";
+#endif
+    constexpr auto char3 = long_chars_sv[3];
+    constexpr auto char3csc = csc(char3);
+    static_assert(char3csc == 195u);
+    constexpr auto char13 = long_chars_sv[13];
+    constexpr auto char13csc = csc(char13);
+    static_assert(char13csc == 172u);
+    constexpr auto char24 = long_chars_sv[24];
+    constexpr auto char24csc = csc(char24);
+    static_assert(char24csc == 32u);
+    static_assert(long_chars_sv.size() == 34);
+    constexpr auto long_chars_csc = []() {
+        uint32_t sum = 0;
+        CheckSumCombine(sum, long_chars_sv);
+        return sum;
+    }();
+    static_assert(long_chars_csc == 5816);
 
 
     static_assert(CHECKSUM_MODULUS < std::numeric_limits<uint32_t>::max());

--- a/util/CheckSums.h
+++ b/util/CheckSums.h
@@ -21,7 +21,7 @@ namespace CheckSums {
 
     // integeral types
     template <typename T> requires (std::is_signed_v<T>)
-    constexpr void CheckSumCombine(uint32_t& sum, T t)
+    constexpr void CheckSumCombine(uint32_t& sum, T t) noexcept
     {
         sum += static_cast<uint32_t>(t >= 0 ? t : -t);
         sum %= CHECKSUM_MODULUS;
@@ -33,20 +33,38 @@ namespace CheckSums {
         sum += static_cast<uint32_t>(t);
         sum %= CHECKSUM_MODULUS;
     }
+    template <>
+    constexpr void CheckSumCombine(uint32_t& sum, signed char t) noexcept
+    {
+        static_assert(noexcept(sum + static_cast<uint32_t>(static_cast<unsigned char>(t))));
+        static_assert(noexcept(sum % CHECKSUM_MODULUS));
+        sum += static_cast<uint32_t>(static_cast<unsigned char>(t));
+        sum %= CHECKSUM_MODULUS;
+    }
+    template <>
+    constexpr void CheckSumCombine(uint32_t& sum, char t) noexcept
+    {
+        if constexpr (std::is_signed_v<char>)
+            CheckSumCombine<signed char>(sum, t);
+        else
+            CheckSumCombine<unsigned char>(sum, t);
+    }
+
 
     // strings
-    constexpr inline void CheckSumCombine(uint32_t& sum, std::string_view sv) {
+    constexpr inline void CheckSumCombine(uint32_t& sum, std::string_view sv) noexcept {
         for (auto t : sv)
             CheckSumCombine(sum, t);
         sum += sv.size();
         sum %= CHECKSUM_MODULUS;
     }
-    constexpr inline void CheckSumCombine(uint32_t& sum, const char* s) { CheckSumCombine(sum, std::string_view{s}); }
+    constexpr inline void CheckSumCombine(uint32_t& sum, const char* s) noexcept
+    { CheckSumCombine(sum, std::string_view{s}); }
 
     // classes that have GetCheckSum methods
     template <typename C> //requires (requires(C c) { c.GetCheckSum(); })
     void CheckSumCombine(uint32_t& sum, const C& c,
-                         decltype(std::declval<C>().GetCheckSum())* = nullptr)
+                         decltype(std::declval<C>().GetCheckSum())* = nullptr) noexcept(noexcept(c.GetCheckSum()))
     {
         sum += c.GetCheckSum();
         sum %= CHECKSUM_MODULUS;
@@ -54,12 +72,12 @@ namespace CheckSums {
 
     // enums
     template <typename T> requires (std::is_enum_v<T>)
-    constexpr void CheckSumCombine(uint32_t& sum, T t)
+    constexpr void CheckSumCombine(uint32_t& sum, T t) noexcept
     { CheckSumCombine(sum, static_cast<int>(t) + 10); }
 
     // pointer types
     template <typename T>
-    constexpr void CheckSumCombine(uint32_t& sum, const T* p)
+    constexpr void CheckSumCombine(uint32_t& sum, const T* p) noexcept
     {
         if (p)
             CheckSumCombine(sum, *p);

